internal/proxy/proxy.go

```diff
@@ -28,6 +29,7 @@ const (
 	CookieJenkinsIdled = "JenkinsIdled"
 	ServiceName        = "jenkins"
 	SessionCookie      = "JSESSIONID"
+	RequestTimeout     = time.Second * 5
```

Comment:
Let me see whether I understand this correctly. Client makes a request for the proxy with a certain timeout (I think Firefox uses something around 30 seconds or so). Our code will try to forward to the underlying Jenkins instance. To do this I makes a request to the Jenkins login page to make sure the Jenkins route is accessible (needed because the Jenkins pod might be running, but HAProxy has not updated the routes yet).

What happens is that the request to the login page actually does not return within the timeout limit of the original request, the original request times out with a 504 error.  

This pull request will change the timeout of the request to Jenkins to be quite low (5 seconds), so it will for sure timeout earlier than the original client request. But what does this mean for the user. He goes from error 504 to a error 500?

Reply:
Hmm..yeah, this PR fixes the "loading page" case, but potentially breaks the actual login case

Reply:
So we need to think more about this again?

Reply:
thinking about it more - this  timeout change does not really change current behaviour - if there is problem with reaching Jenkins at the moment, there will be timeout anyway - in this case it only does not take 30s, but you get error after 5s.

There might be better way to handle things, I guess - add retries somewhere? Handle login error by redirect rather then returning error message?

Reply:
> thinking about it more - this timeout change does not really change current behaviour - if there is problem with reaching Jenkins at the moment, there will be timeout anyway 

That was my thinking. You might change the error here 500 vs 504, but I am not sure. I have even not seen the 504 error yet, but admittingly I have not tried very hard either. I would like to get into a position where we can test the HTTP flow independent of the actual logic happening. On a very high coarse level, we could introduce a GitHubRequestHandler and a JenkinsRequestHandler (as interfaces). Proxy will be instantiated with implementations of these interfaces matching the current implementation. Proxy.Handle calls then just these interfaces depending on the type of request. Having this interface allows us to plug in dummy implementations of these handlers which don't call any external dependencies, but rather mock certain conditions which can occur. This would take some doing, but I don't see really any other way to test and fully understand what's going on.

> There might be better way to handle things, I guess - add retries somewhere? Handle login error by redirect rather then returning error message?

Probably, but I think we need to gain a better understanding first on what's going on. For now this seems a bit premature. I would for now probably close this pull request, but leave the issue itself open. As mentioned above I would start with some more code changes in order to get into a position where I can reproduce certain conditions.